src/model.rs

```diff
@@ -47,45 +49,115 @@ impl Marginal {
-                            .to_vec()
-                            .iter()
-                            .zip(fractions[0..haplotype_index].to_vec().iter())
-                        {
-                            if (h == haplotype_query) && (f > &NotNan::new(0.0).unwrap()) {
+                        //loop over haplotypes in the distance matrix and find haplotypes that are at distance x to the current haplotype (similar_L(h))
+                        if let Some(distance_matrix) = &self.distance_matrix {
+                            dbg!(&distance_matrix);
+                            dbg!(&self.lp_haplotypes);
+
+                            let mut lp_haplotype_list = self.lp_haplotypes.clone().unwrap();
+
+                            //retain all elements in lp haplotypes except the current haplotype
+                            lp_haplotype_list.retain(|x| x != current_haplotype);
+
+                            //compute similar_L(current_haplotype)
+                            let mut similar_l = 0;
+
+                            for h in lp_haplotype_list.iter() {
+                                for ((h1, h2), distance) in distance_matrix.iter() {
+                                    if ((h1 == h && h2 == current_haplotype)
+                                        || (h2 == h && h1 == current_haplotype))
+                                        && (*distance < x)
+                                    {
+                                        similar_l += 1;
+                                    }
+                                }
+                            }
+                            dbg!(&similar_l);
+                            dbg!(&fractions);
+
+                            //loop over haplotypes in collected fractions and find haplotypes that are at distance x to the current haplotype (similar_R(current_haplotype))
+                            let mut similar_r = 0;
+                            for (h, f) in self.haplotypes[0..haplotype_index]
+                                .to_vec()
+                                .iter()
+                                .zip(fractions[0..haplotype_index].to_vec().iter())
+                            {
+                                //for each collected fraction, loop over distance matrix and check if there is a haplotype with fraction > 0.0
+                                for ((h1, h2), distance) in distance_matrix.iter() {
+                                    if ((h1 == h && h2 == current_haplotype)
+                                        || (h2 == h && h1 == current_haplotype))
+                                        && (*distance < x)
+                                        && (f > &NotNan::new(0.0).unwrap())
+                                    {
+                                        dbg!(&h, &f, &h1, &h2);
+                                        similar_r += 1;
+                                    }
+                                }
+                            }
```

Comment:
Again, this is very slow inside of the recursion. Maybe the same kind of datastructure as above can be reused for this here.

Reply:
Similar_L(h) is already known before the model hence it can be calculated and we can do lookups. But isn't Similar_R(h) dynamically determined for each recursion arm?

Reply:
Yes, but the distance matrix access will always iterate over the entire distance matrix instead of doing a constant time lookup for (h, current_haplotype). The latter would be much faster.